### PR TITLE
migrate to ppxlib

### LINKED ISSUE #141: Port ppx_distr_guards to ppxlib
https://github.com/vogler/ppx_distr_guards

### PR DESCRIPTION
**Pull request by tatchi**
Fixes https://github.com/ocaml-ppx/ppxlib/issues/141

### CODE REVIEW

File: dune-project [hunk 1]
```diff
@@ -15,7 +15,7 @@
   (description "`function%distr A x, _ | _, A x when p x -> e` will result in `function A x, _ when p x -> e | _, A x when p x -> e`")
   (depends
    (dune (>= 2.3.0))
-   (ocaml-migrate-parsetree (>= 1.6.0))
+   ppxlib
```
Comment (tatchi) [inline dune-project]: Should we specify any version constraint?

File: ppx_distr_guards.ml [hunk 2]
```diff
@@ -1,44 +1,39 @@
...
+and match_expand ~ctxt expr cases =
+  let new_cases = cases |> List.map (case_to_cases ~ctxt) |> List.flatten in
+  let loc = Expansion_context.Extension.extension_point_loc ctxt in
+  Ast_helper.Exp.match_ ~loc expr new_cases
+
+and expand ~ctxt e =
+  match e.pexp_desc with
+  | Pexp_function cases -> function_expand ~ctxt cases
+  | Pexp_match (e, cases) -> match_expand ~ctxt e cases
+  | _ -> e
```
Comment (pitag-ha) [inline ppx_distr_guards.ml]: Just one doubt: what does the expansion look like for a `Pexp_try` payload, such as
```
exception A of int
exception B of int

let _ =  [%distr try <whatever> with | A n | B n when n <> 0 -> 1
```
before and after the port?

File: dune-project (hunk 1 above)
Comment (pitag-ha) [reply dune-project]: I don't think that would be necessary. A lot of ppx rewriter folks like to avoid a dependency on `base` and `stdio` though. For that you'd need `ppxlib >= 0.15.0`.

**Comment by pitag-ha** [comment]
Thanks (from the ppxlib side) for the port!

File: ppx_distr_guards.ml [hunk 3]
```diff
@@ -1,44 +1,39 @@
...
-      | x -> default_mapper.expr m x;
-  }
-
-let () =
-  Driver.register ~name:"distr_guards" ocaml_version mapper
-
-(*
-ocamlc -dparsetree test.ml
-ocamlbuild -package compiler-libs.common ppx_distr_guards.native
-ocamlc -dsource -ppx ./ppx_distr_guards.native test.ml
```
Comment (vogler) [inline ppx_distr_guards.ml:43]: Can you keep those and adjust them if needed?
I wanted to have this repo as a template for ppx extensions and it's nice to have everything needed for development/test in one file.

File: ppx_distr_guards.ml [hunk 4]
```diff
@@ -1,44 +1,39 @@
...
+      let rec f a p =
+        match p.ppat_desc with Ppat_or (p1, p2) -> f (f a p2) p1 | _ -> p :: a
+      in
+      let ps = f [] pc_lhs in
+      List.map (fun p -> Ast_helper.Exp.case p ~guard pc_rhs) ps
+  | x -> [ { x with pc_rhs = expand ~ctxt x.pc_rhs } ]
+
+and function_expand ~ctxt cases =
+  let new_cases = cases |> List.map (case_to_cases ~ctxt) |> List.flatten in
+
```
Comment (vogler) [inline ppx_distr_guards.ml]: Why the empty line?

File: ppx_distr_guards.ml (hunk 2 above)
Comment (vogler) [reply ppx_distr_guards.ml]: Haven't checked out `ppxlib` but this expand duplication looks pretty ugly/error-prone compared to the visitor with `cases`.
It doesn't have anything to abstract this?

**Review by vogler — CHANGES_REQUESTED** [review]

File: ppx_distr_guards.ml (hunk 2 above)
Comment (tatchi) [reply ppx_distr_guards.ml]: > Just one doubt: what does the expansion look like for a `Pexp_try` payload, such as
> 
> ```
> exception A of int
> exception B of int
> 
> let _ =  [%distr try <whatever> with | A n | B n when n <> 0 -> 1
> ```
> 
> before and after the port?

**Before the port:**

```ocaml
exception A of int 
exception B of int 
let _ = try Sys.readdir "" with | A n when n <> 0 -> 1 | B n when n <> 0 -> 1
```

**After the port:**

```ocaml
exception A of int 
exception B of int 
let _ = try Sys.readdir "" with | A n|B n when n <> 0 -> 1
```

Good catch! I will try to cover that case and add a test for it. Can you think of something else that might be missing?

File: ppx_distr_guards.ml (hunk 3 above)
Comment (tatchi) [reply ppx_distr_guards.ml:43]: Oops sorry. What's the purpose of the two last commands?

```
ocamlbuild -package compiler-libs.common ppx_distr_guards.native
ocamlc -dsource -ppx ./ppx_distr_guards.native test.ml
```

if it's used to print the transformed code, I've personally used:

```
./_build/default/standalone.exe --impl test.ml
```

**Comment by tatchi** [comment]
> Haven't checked out `ppxlib` but this expand duplication looks pretty ugly/error-prone compared to the visitor with `cases`.
> It doesn't have anything to abstract this?

@pitag-ha any idea to improve that?

File: ppx_distr_guards.ml (hunk 3 above)
Comment (vogler) [reply ppx_distr_guards.ml:43]: Yes, how to build as standalone and transform code.
Also, just to list an alternative to dune if it's not installed or one does not want to depend on the `dune` file.
Does `ocamlbuild -package ppxlib ppx_distr_guards.native` work?

Comment (tatchi) [reply ppx_distr_guards.ml:43]: > Does `ocamlbuild -package ppxlib ppx_distr_guards.native` work?

Yes. At least a `ppx_distr_guards.native` file is created. But when I run

```
ocamlc -dsource -ppx ./ppx_distr_guards.native test.ml
```

I get this error: `Error: External preprocessor does not produce a valid file`

**Comment by pitag-ha** [comment]
> Haven't checked out ppxlib but this expand duplication looks pretty ugly/error-prone compared to the visitor with cases.
It doesn't have anything to abstract this?

It's not possible to do the same as before here, since the ppx is (and, I think, should be) written as a context-free rule now. That has the advantage that when using this ppx, it will be applied together with all other registered context-free rules in one single AST pass. To be able to do only one single AST pass, ppxlib keeps the AST recursion internal and extension node rewriters only have access to the extension node itself.

Usually there is a way for context-free rules to abstract the the expected pattern inside the payload: a more complex [`Ast_pattern`](https://ocaml-ppx.github.io/ppxlib/ppxlib/Ppxlib/Ast_pattern/index.html) than `single_expr_payload` as used here. The problem is that here, when receiving a different pattern than the expected one, you don't want to raise as usual but want to attach the whole payload as is to the node. So making better use of `Ast_pattern` here as usual is not an option.

Also, another typical way to make this less verbose would be to use [metaquot](https://ppxlib.readthedocs.io/en/latest/ppx-for-plugin-authors.html#metaquot). But there's no way to anti-quote `cases`. So again, here that's not an option.

So, long story short: I can't think of a better way in this particular case. To look on the bright site, making the pattern explicit also seems to have an advantage: for example, before it was easy to forget that a `Pexp_try` payload is also captured by the pattern, whereas now that's explicit.

File: ppx_distr_guards.ml (hunk 2 above)
Comment (pitag-ha) [reply ppx_distr_guards.ml]: > Can you think of something else that might be missing?

It's all covered now. There are only three nodes that hold `cases` and you're matching them all now.

File: ppx_distr_guards.ml (hunk 3 above)
Comment (tatchi) [reply ppx_distr_guards.ml:43]:
```
ocamlfind ocamlc ppx_distr_guards.ml -package ppxlib -linkpkg standalone.ml -o standalone.exe
./standalone.exe --impl test.ml
```

That seems to work, but I'm definitely not sure if that's the best way of doing it 😅

Comment (pitag-ha) [reply ppx_distr_guards.ml:43]: I don't know much about `ocamlfind` and `ocamlc`, that's also why I haven't answered to this yet (but you basically want to compile `ppx_distr_guards.ml` and `standalone.ml`, in that order, linking in `ppxlib`. Is that what you do?). But I do agree that using the ppx driver directly as you do in `./standalone.exe --impl test.ml` is definitely the right way to go nowadays.

**Comment by pitag-ha** [comment]
@vogler, is there anything blocking this from being merged and released?

**Comment by vogler** [comment]
Sorry, busy with other stuff... Wanted to have a look at ppxlib and abstract the cases code. I'll try to get to it this weekend.

**Comment by vogler** [comment]
Ok, did what I wanted to do. Test passes - I hope it otherwise still behaves the same. 😃
@pitag-ha I didn't quite get the details of your comment above. With the last commit there's less duplicated code. I also see no problem with having a `map_cases` similar to the new `expand` in ppxlib.

**Comment by pitag-ha** [comment]
Nice!

> I didn't quite get the details of your comment above

I understood that you wanted an abstraction with objects as it was before with OMP.

**Review by vogler — APPROVED** [review]

**Comment by vogler** [comment]
> I understood that you wanted an abstraction with objects as it was before with OMP.

No, just any abstraction :)

Opam-CI is green, so should be available soon.
